File: packages/editor/src/display-id.ts
// Adapted from packages/contracts/src/display-id.ts at source SHA
// 393795261322b916e588043cf94feca999175843.

/*
 * WHY: #807 B2 — `KEY-n`/`WIKI-n` 문법 정본. href·lookupDisplayId 와 같은 모양이며
 * 접두는 projectKeyPattern(2–32, 하이픈)과 맞춘다. 번호는 9자리로 잘라 정수 범위 안.
 */
export const displayIdPattern = /^([A-Za-z0-9-]{2,32})-(\d{1,9})$/;

export type ParsedDisplayId = { prefix: string; n: number };

export function parseDisplayId(raw: string): ParsedDisplayId | null {
	const match = displayIdPattern.exec(raw.trim());
	const prefix = match?.[1];
	const digits = match?.[2];
	if (prefix === undefined || digits === undefined) return null;
	return { prefix: prefix.toUpperCase(), n: Number(digits) };
}

export function formatDisplayId(prefix: string, n: number): string {
	return `${prefix}-${n}`;
}

File: src/display_id.rs
use regex::Regex;
use std::sync::LazyLock;

static DISPLAY_ID_RE: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r"^([A-Za-z0-9-]{2,32})-(\d{1,9})$").expect("display id regex"));

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ParsedDisplayId {
    pub prefix: String,
    pub number: i32,
}

pub fn parse_display_id(raw: &str) -> Option<ParsedDisplayId> {
    let trimmed = raw.trim();
    let caps = DISPLAY_ID_RE.captures(trimmed)?;
    let prefix = caps.get(1)?.as_str().to_uppercase();
    let digits = caps.get(2)?.as_str();
    let number = digits.parse().ok()?;
    Some(ParsedDisplayId { prefix, number })
}

pub fn format_display_id(prefix: &str, number: i32) -> String {
    format!("{prefix}-{number}")
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parses_key_and_wiki_display_ids() {
        assert_eq!(
            parse_display_id("lab-12"),
            Some(ParsedDisplayId {
                prefix: "LAB".to_string(),
                number: 12
            })
        );
        assert_eq!(
            parse_display_id(" WIKI-3 "),
            Some(ParsedDisplayId {
                prefix: "WIKI".to_string(),
                number: 3
            })
        );
        assert!(parse_display_id("x").is_none());
    }
}
